### PR TITLE
add utility library for type-level sized vectors

### PR DESCRIPTION
**Pull request by fricklerhandwerk**
also a few other things on the side

### CODE REVIEW

**Review by gschult — APPROVED** [review]

**Review by gschult — APPROVED** [review]

**Review by gschult — APPROVED** [review]

File: src/FixedVector/TH.hs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+--       createVector<N> v1 v2 {...} v<N> = prepend v1 $ createVector<N-1> v2 ... v<N>
+createVectorN :: Natural -> DecsQ
+createVectorN n =
+  pure
+    [ SigD functionName functionType
+    , FunD functionName clauses
+    ]
+ where
+  -- here we have access to n from `createVector``
+  -- the name of the function we are creating
```
Comment (fricklerhandwerk) [inline src/FixedVector/TH.hs]:
```suggestion
```

The comments don't explain anything, and the declaration appears trivial.

File: src/FixedVector/TH.hs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+import FixedVector.Type
+import Language.Haskell.TH
+import Numeric.Natural (Natural)
+
+-- import Language.Haskell.TH.Lib
+
+-- create a function that we would have written as follows:
+-- createVector<N> :: t -> t -> {N times} -> t -> FixedVector t <N>
+-- createVector<N> v1 v2 {...} v<N> = FixedVector { _data = [ v1, v2, {...}, v<N> ] }
```
Comment (fricklerhandwerk) [inline src/FixedVector/TH.hs]:
```suggestion
-- createVector<N> :: t -> t -> ... -> t -> FixedVector t <N>
-- createVector<N> v1 v2 ... v<N> = FixedVector { _data = [ v1, v2, ..., v<N> ] }
```

For better readability.

File: src/FixedVector/TH.hs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+-- using Template Haskell: https://hackage.haskell.org/package/template-haskell-2.19.0.0/docs/Language-Haskell-TH.html
+-- TODO: don't depend on internals of FixedVector and instead use exposed constructors
+--       this might only be possible once we have e.g. createVectorUpToN, because then we can depend on the existence of these constructors
```
Comment (fricklerhandwerk) [inline src/FixedVector/TH.hs:16]:
```suggestion
```

@ners will TH recursively declare `createVector<N-1>` when we call `createVector N`?

File: src/FixedVector/TH.hs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+--       createVector<N> v1 v2 {...} v<N> = prepend v1 $ createVector<N-1> v2 ... v<N>
```
Comment (fricklerhandwerk) [inline src/FixedVector/TH.hs:17]:
```suggestion
--       createVector<N> v1 v2 ... v<N> = prepend v1 $ createVector<N-1> v2 ... v<N>
```

For consistency and readability

File: src/FixedVector/TH.hs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  functionName :: Name
+  functionName = mkName $ "createVector" ++ show n
+
+  -- The type of elements stored in our vector, which we denote as `t`
+  vectorElementType :: Type
+  vectorElementType = VarT $ mkName "t"
+
+  -- ''FixedVector is the Name of the concrete type FixedVector
+  returnType :: Type
+  returnType = ConT ''FixedVector `AppT` vectorElementType `AppT` LitT (NumTyLit $ fromIntegral n)
```
Comment (fricklerhandwerk) [inline src/FixedVector/TH.hs]: Constructing the value is IMHO non-trivial, so we should explain what it's supposed to produce.
```suggestion
  -- FixedVector t <N>
  returnType :: Type
  -- ''FixedVector is the Name of the concrete type FixedVector
  returnType = ConT ''FixedVector `AppT` vectorElementType `AppT` LitT (NumTyLit $ fromIntegral n)
```

Add the note on notation to the implementation, not the overall declaration.

File: src/FixedVector/TH.hs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  -- alternatively, equivalent:
+  -- returnType = AppT (AppT (ConT ''FixedVector) vectorElementType) (LitT $ NumTyLit n)
+  -- returnType = foldr1 (AppT) [ConT ''FixedVector, vectorElementType, LitT (NumTyLit n)]
+
+  -- create the following expression:
+  -- t -> t -> ... -> t -> t -> FixedVector t n
+  -- in terms of our intermediate functions:
+  -- vectorElementType -> vectorElementType -> ... -> returnType
```
Comment (fricklerhandwerk) [inline src/FixedVector/TH.hs]:
```suggestion
  -- t -> t -> ... -> t -> t -> FixedVector t <N>
```

We made the implementation much more readable, so we can reduce the amount of explanations to what we expect to see as a result.

File: src/FixedVector/TH.hs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+  -- Documentation:
+  clauses :: [Clause]
+  clauses = [Clause pats body decls]
+   where
+    -- function parameters
+    -- v1 v2 ... v<N>
+    pats = VarP <$> vars
+    -- results in:
+    -- FixedVector { _data = [ v1, v2, ..., v<N> ] }
```
Comment (fricklerhandwerk) [inline src/FixedVector/TH.hs]:
```suggestion
    -- v1 v2 ... v<N>
    pats = VarP <$> vars
    -- FixedVector { _data = [ v1, v2, ..., v<N> ] }
```

File: src/FixedVector/TH.hs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+   where
+    -- 0th :: returnType
+    -- 1st :: AppT vectorElementType (AppT ArrowT returnType)
+    -- 2nd :: AppT vectorElementType (AppT ArrowT (AppT (vectorElementType) (AppT ArrowT returnType)))
+    -- 3rd :: AppT vectorElementType (AppT ArrowT (AppT vectorElementType (AppT ArrowT (AppT (vectorElementType) (AppT ArrowT returnType)))))
+    -- ... (really boring)
+    -- nth is what we need!
+    prependTerm t u = ArrowT `AppT` t `AppT` u
+
+  -- Documentation:
```
Comment (fricklerhandwerk) [inline src/FixedVector/TH.hs]:
```suggestion
  -- v1 v2 ... v<N> = FixedVector { _data = [ v1, v2, ..., v<N> ] }
```

File: src/FixedVector/TH.hs [hunk 9]
```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module FixedVector.TH where
+
+import FixedVector.Type
+import Language.Haskell.TH
+import Numeric.Natural (Natural)
+
+-- import Language.Haskell.TH.Lib
```
Comment (fricklerhandwerk) [inline src/FixedVector/TH.hs:9]:
```suggestion
```

File: src/FixedVector/TH.hs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    -- function parameters
+    -- v1 v2 ... v<N>
+    pats = VarP <$> vars
+    -- results in:
+    -- FixedVector { _data = [ v1, v2, ..., v<N> ] }
+    body = NormalB $ RecConE (mkName "FixedVector") [('_data, ListE $ VarE <$> vars)]
+    -- `where` clauses, which we currently don't use
+    decls = []
+    vars :: [Name]
+    vars = mkName . ("v" ++) . show <$> [1 .. n]
```
Comment (fricklerhandwerk) [inline src/FixedVector/TH.hs]:
```suggestion
    vars :: [Name]
    vars = mkName . ("v" ++) . show <$> [1 .. n]
    -- `where` clauses, which we currently don't use
    decls = []
```

Declarations can move to the end, we don't have to look at them before anything else. This is purely informative for the reader.

PS: @ners I still think that reading at least some chunks linearly is a legitimate use case for any codebase.

**Review by fricklerhandwerk — COMMENTED** [review]
This still needs a proper commit message.

As a side note, I think we went overboard with the strong type-level assurance. While I learned a great deal about Template Haskell, it may have been more productive to play with the actual logic on stupid lists first, and then migrate towards more type safety once the need for it arises.

Really looking forward to wrangle some bytes.

File: src/FixedVector/TH.hs (hunk 5 above)
Comment (gschult) [reply src/FixedVector/TH.hs]: I like your reminder on the type we expect to receive.

File: src/FixedVector/TH.hs (hunk 8 above)
Comment (gschult) [reply src/FixedVector/TH.hs]: This looks like a bad function definition to me. I find the explanations inside the clauses more helpful.

File: src/FixedVector/TH.hs (hunk 4 above)
Comment (gschult) [inline src/FixedVector/TH.hs:17]: We should drop this one. This is not how we implemented createVector<N> in TH.

File: src/FixedVector/TH.hs (hunk 3 above)
Comment (ners) [reply src/FixedVector/TH.hs:16]: No. `createVector N` only declares `createVectorN`. That is why I suggest adding utilities such as `createVectorUpTo`.

**Pull request merged**